### PR TITLE
Add script for rebasing onto later OKD version

### LINKED ISSUE #232: Add script to rebase onto a new OKD version
Create a `scripts/rebase.sh` script that takes as parameter the OKD release version string to rebase to and then updates
* vendored packages to the ones used by that OKD release,
* image digests in `pkg/release/release_*.go` for architectures supported by OKD,
* manifests in the assets directory

based on the release image metadata.

#### Why is this needed:
Minimize manual processing of rebasing to a later OKD version. Should hopefully lead to more frequent rebases.

### PR DESCRIPTION
**Pull request by fzdarsky**
This patch adds a scripts/rebase.sh script that takes as parameter the OKD release version string to rebase to and then updates
* vendored packages to the ones used by that OKD release,
* image digests in pkg/release/release_*.go for architectures supported by OKD,
* manifests in the assets directory
based on the release image metadata.

Signed-off-by: Frank A. Zdarsky <fzdarsky@redhat.com>

Closes #232

### CODE REVIEW

File: scripts/rebase.sh [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+)
+EOF
+
+go mod tidy
+go mod vendor
+
+
+title "Rebasing release_*.go"
+for arch in amd64; do
+    images="$(get_release_images ${REPOROOT}/pkg/release/release_${arch}.go | xargs)"
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply scripts/rebase.sh:188]: that ${REPOROOT} may need also quotes, but... not sure what would work exactly here...
```suggestion
    images=$(get_release_images "${REPOROOT}/pkg/release/release_${arch}.go" | xargs)
```

may be this? hmm

File: scripts/rebase.sh [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+title "Rebasing manifests"
+assets=$(find ${REPOROOT}/assets -name "*.yaml")
+for asset in ${assets}; do
+    search_path=${REPOROOT}/_output/staging/release-manifests
+    search_name=$(basename ${asset})
+    search_exclude=XXX
+
+    # TODO: Rename assets and their references to obviate the need for special cases
+    case $(basename ${asset}) in
```
Comment (mangelajo) [inline scripts/rebase.sh:209]:
```suggestion
    case $(basename "${asset}") in
```

File: scripts/rebase.sh [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        search_name="${search_name#"0000_80_openshift-router-"}"
+        search_exclude="${REPOROOT}/_output/staging/cluster-dns-operator/assets/router/metrics/*"
+        ;;
+    0000_11_imageregistry-configs.crd.yaml)
+        search_path=${REPOROOT}/_output/staging/openshift-apiserver/vendor/github.com/openshift/api/imageregistry/v1
+        search_name=00-crd.yaml
+        ;;
+    esac
+
+    updated_asset=$(find ${search_path} -name "${search_name}" -not -path "${search_exclude}" | tail -n 1)
```
Comment (mangelajo) [inline scripts/rebase.sh:252]:
```suggestion
    updated_asset=$(find "${search_path}" -name "${search_name}" -not -path "${search_exclude}" | tail -n 1)
```

**Review by mangelajo — CHANGES_REQUESTED** [review]
Some nits, and may be some useful ones ':D

File: scripts/rebase.sh [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+title() {
+    echo -e "\E[34m\n$1\E[00m";
+}
+
+# Reads go.mod file $1 and prints lines in its section $2 ("require" or "replace")
+extract_section() {
+    file=$1
+    section=$2
+
+    cat ${file} | awk "BEGIN {output=0} /^)/ {output=0} {if (output == 1 && !match(\$0,\"// indirect\")) print \$0} /^${section}/ {output=1}"
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: added, thx

File: scripts/rebase.sh [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        if [[ "${line}" =~ ^//.* ]]; then
+            continue
+        fi
+
+        mod="$(get_mod "${line}")"
+        base_version=$(get_version "${line}")
+        version=${base_version}
+
+        update_line=$(cat "${update_file}" | grep "${mod}" || true)
+        if [[ ! -z "${update_line}" ]]; then
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: added, thx

File: scripts/rebase.sh [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            update_version=$(get_version "${update_line}")
+            version=$(printf '%s\n%s\n' "${base_version}" "${update_version}" | sort --version-sort | tail -n 1)
+        fi
+
+        echo "${mod} ${version}" 
+    done < ${base_file}
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: added, thx

File: scripts/rebase.sh [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+# Returns the list of release image names from a release_${arch}.go file
+get_release_images() {
+    file=$1
+
+    cat ${file} | awk "BEGIN {output=0} /^}/ {output=0} {if (output == 1) print substr(\$1, 2, length(\$1)-3)} /^var Image/ {output=1}"
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: added, thx

File: scripts/rebase.sh [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+
+# == MAIN ==
+if [[ $EUID -ne 0 ]]; then
+   >&2 echo "You need to run this script as root or in a `buildah unshare` environment:"
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: yeah, bad markdown habit ;P

File: scripts/rebase.sh [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+
+title "Rebasing go.mod..."
+extract_section ${REPOROOT}/go.mod require > latest_require
+extract_section ${REPOROOT}/go.mod replace > latest_replace
+while IFS="" read -r line || [ -n "$line" ]
+do
+    COMPONENT=$(echo "${line}" | cut -d ' ' -f 1)
+    REPO=$(echo "${line}" | cut -d ' ' -f 2)
+    if [[ ${EMBEDDED_COMPONENTS} == *"${COMPONENT}"* ]]; then
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: added, thx

File: scripts/rebase.sh (hunk 2 above)
Comment (fzdarsky) [reply scripts/rebase.sh]: fixed

File: scripts/rebase.sh (hunk 3 above)
Comment (fzdarsky) [reply scripts/rebase.sh]: fixed

File: scripts/rebase.sh [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+        extract_section ${REPO##*/}/go.mod require > require
+        extract_section ${REPO##*/}/go.mod replace > replace
+        update_versions latest_require require > t; mv t latest_require
+        update_versions latest_replace replace > t; mv t latest_replace
+    fi
+done < source_commits.txt
+
+cat << EOF > ${REPOROOT}/go.mod
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: fixed

File: scripts/rebase.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (mangelajo) [reply scripts/rebase.sh:194]: I'm not sure if this still needs to be fixed, ^ @fzdarsky , probably I'm missing something, approving but have an eye before merge.

**Review by mangelajo — APPROVED** [review]

File: scripts/rebase.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: Found a combo that should work and updated.

File: scripts/rebase.sh [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+
+
+title "Downloading and extracting ${OKD_RELEASE} release image..."
+curl -LO "https://github.com/openshift/okd/releases/download/${OKD_RELEASE}/release.txt"
+
+OKD_RELEASE_IMAGE=$(grep -oP 'Pull From: \K[\w.-/@:]+' release.txt)
+podman pull ${OKD_RELEASE_IMAGE}
+cnt=$(buildah from ${OKD_RELEASE_IMAGE})
+mnt=$(buildah mount ${cnt} | cut -d ' ' -f 2)
+cat ${mnt}/release-manifests/image-references \
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply scripts/rebase.sh]: [x] done

**Pull request merged**